### PR TITLE
refactor: refactor drop share

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. refactor: refactor drop share
2. remove db directly if created from a share.
3. add more test case.

Fixes #issue

### CODE REVIEW

File: src/meta/api/src/kv_api_utils.rs [hunk 1]
```diff
@@ -414,3 +418,91 @@ pub fn get_share_database_id_and_privilege(
         ShareHasNoGrantedDatabase::new(&name_key.tenant, &name_key.share_name),
     )))
 }
+
+// Return true if all the database data has been removed.
+pub async fn is_all_db_data_has_been_removed(
```
Comment (drmingdrmer) [inline src/meta/api/src/kv_api_utils.rs:423]:
```suggestion
pub async fn is_all_db_data_removed(
```

File: src/meta/api/src/kv_api_utils.rs [hunk 2]
```diff
@@ -414,3 +418,91 @@ pub fn get_share_database_id_and_privilege(
...
+    kv_api: &(impl KVApi + ?Sized),
+    db_id: u64,
+) -> Result<bool, MetaError> {
+    let dbid = DatabaseId { db_id };
+
+    let (db_meta_seq, db_meta): (_, Option<DatabaseMeta>) = get_struct_value(kv_api, &dbid).await?;
+    if db_meta_seq != 0 || db_meta.is_some() {
+        return Ok(false);
+    }
```
Comment (drmingdrmer) [inline src/meta/api/src/kv_api_utils.rs:432]: Checking `db_meta_seq` would be enough. `db_meta_seq == 0` and `db_meta` is `Some` is a bug.

```suggestion
    debug_assert_eq!((db_meta_seq == 0), db_meta.is_none());
    if db_meta_seq != 0 {
        return Ok(false);
    }
```

File: src/meta/api/src/kv_api_utils.rs [hunk 3]
```diff
@@ -414,3 +418,91 @@ pub fn get_share_database_id_and_privilege(
...
+
+    let id_to_name = DatabaseIdToName { db_id };
+    let (name_ident_seq, name_ident): (_, Option<DatabaseNameIdent>) =
+        get_struct_value(kv_api, &id_to_name).await?;
+    if name_ident_seq != 0 || name_ident.is_some() {
+        return Ok(false);
+    }
```
Comment (drmingdrmer) [inline src/meta/api/src/kv_api_utils.rs:439]: ditto

File: src/meta/api/src/kv_api_utils.rs [hunk 4]
```diff
@@ -414,3 +418,91 @@ pub fn get_share_database_id_and_privilege(
...
+) -> Result<(bool, Option<ShareNameIdent>), MetaError>
+where
+    F: FnMut(&DatabaseMeta) -> bool,
+{
+    let dbid = DatabaseId { db_id };
+
+    let (db_meta_seq, db_meta): (_, Option<DatabaseMeta>) = get_struct_value(kv_api, &dbid).await?;
+    if db_meta_seq == 0 || db_meta.is_none() {
+        return Ok((false, None));
+    }
```
Comment (drmingdrmer) [inline src/meta/api/src/kv_api_utils.rs:459]: ditto

File: src/meta/api/src/kv_api_utils.rs [hunk 5]
```diff
@@ -414,3 +418,91 @@ pub fn get_share_database_id_and_privilege(
...
+
+    let id_to_name = DatabaseIdToName { db_id };
+    let (name_ident_seq, name_ident): (_, Option<DatabaseNameIdent>) =
+        get_struct_value(kv_api, &id_to_name).await?;
+    if name_ident_seq == 0 || name_ident.is_none() {
+        return Ok((false, None));
+    }
```
Comment (drmingdrmer) [inline src/meta/api/src/kv_api_utils.rs:466]: ditto

**Review by drmingdrmer — COMMENTED** [review]
Love it

**Review by drmingdrmer — APPROVED** [review]

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**